# Atividade prática 1
### Números de Fibonacci

A sequência de Fibonacci é uma velha conhecidade na matemática.
Originalmente usada para modelar a dinâmica populacional de coelhos em 1202, ela hoje encontra aplicações nos mais diversos campos do conhecimento, como a computação, botânica, arquitetura, etc.

Podemos definir a sequência de Fibonacci $\left(F_n\right)_{n=0}^\infty$ usando recursividade. Com efeito, definimos $F_0 = 0$ e $F_1 = 1$, os quais chamamos de casos base. Agora, estabelecemos a seguinte relação recursiva:

$$F_{n} = F_{n-1} + F_{n-2}\text{, para } n = 2, 3, 4, \ldots$$

Uma implementação recursiva trivial dessa recorrência é a seguinte:


In [1]:
def fibrec(n):
    """
    Algoritmo recursivo para o cálculo dos números da sequência de Fibonacci.
    
    Entrada:
         n (int): número inteiro não negativo.
         
    Saída:
         (int): o valor de F_n
    """
    # Asseguramos que o valor de n é não negativo
    assert n >= 0
    
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fibrec(n-1) + fibrec(n-2)

Experimente alterar o valor de $n$ abaixo e confira o resultado:

In [59]:
%%time
fibrec(32)

CPU times: user 1.26 s, sys: 1.98 ms, total: 1.26 s
Wall time: 1.27 s


2178309

Infelizmente, o algoritmo recursivo está longe de ser o mais rápido para calcular o $n$-ésimo número de Fibonacci. Como já era de se esperar (não é?), tamanha a sua simplicidade.
Anote o tempo total de CPU acima.
Ele será necessário mais na frente.

Agora, iremos conhecer um algoritmo matricial bastante elegante que nos permitirá usar os arranjos da NumPy para calcular os termos dessa sequência de modo mais eficiente.
A ideia chave do algoritmo matricial está em expressar $F_n$ do seguinte modo:

$$
\begin{bmatrix}F_{n}\\ F_{n-1}\end{bmatrix} = 
\begin{bmatrix}F_{n-1} + F_{n-2}\\ F_{n-1} \end{bmatrix} =
\begin{bmatrix}1 & 1\\ 1 & 0\end{bmatrix}
\begin{bmatrix}F_{n-1}\\ F_{n-2}\end{bmatrix}\text{.}
$$

Definindo

$$
\mathbf{M} = \begin{bmatrix}1 & 1\\ 1 & 0\end{bmatrix}\text{,}
$$

teremos:

\begin{align}
\begin{bmatrix}F_{n}\\ F_{n-1}\end{bmatrix} &= \mathbf{M} \begin{bmatrix}F_{n-1}\\ F_{n-2}\end{bmatrix}\text{,}\\
 &= \mathbf{M}^2 \begin{bmatrix}F_{n-2}\\ F_{n-3}\end{bmatrix}\text{,}\\
 &= \mathbf{M}^3 \begin{bmatrix}F_{n-3}\\ F_{n-4}\end{bmatrix}\text{,}\\
 &\vdots \\
 &= \mathbf{M}^{n-1} \begin{bmatrix}F_{1}\\ F_{0}\end{bmatrix} = \mathbf{M}^{n-1} \begin{bmatrix}1\\ 0\end{bmatrix}\text{.}
\end{align}

Supondo que, ao final,

$$
\textbf{M}^{n-1} =
\begin{bmatrix}
m_{00} & m_{01}\\
m_{10} & m_{11}
\end{bmatrix},
$$

então:

$$
\begin{bmatrix}F_{n}\\ F_{n-1}\end{bmatrix}
=
\mathbf{M}^{n-1} \begin{bmatrix}1\\ 0\end{bmatrix}
=
\begin{bmatrix}
m_{00} & m_{01}\\
m_{10} & m_{11}
\end{bmatrix}
\begin{bmatrix}1\\ 0\end{bmatrix}
=
\begin{bmatrix}
m_{00}\\
m_{10}
\end{bmatrix}.
$$

Portanto, $F_n = m_{00}$.

(a) Inicialmente, aplique o que você aprendeu sobre arranjos da numpy para completar a função `matpot` a seguir. Ela deve **retornar** a $k$-ésima potência de uma matriz $\mathbf{A}$, $n\times n$. Você deve realizar o produto de matrizes utilizando a função `np.dot`.

In [3]:
import numpy as np

In [4]:
def matpot(A,k):
    """
    Calcula a k-ésima potência da matriz A.
    
       Argumentos:
         A (numpy.array): matriz quadrada armazenada usando o `array` da numpy.
         k (int): valor da potência desejada, k >= 0.
         
       Retorno:
         (numpy.array): o valor de A^k
    """
    # você deve iniciar a implementação desta função a partir daqui.
    assert k>=0

    An = A
    for i in range(k):
      An = np.dot(A, An)
    return An
    

Durante o desenvolvimento, verifique sua implementação da `matpot` tomando

$$
\mathbf{A} = \begin{bmatrix}
 2 & -2 & -4 \\
-1 &  3 &  4 \\
 1 & -2 & -3
\end{bmatrix}
\qquad
e
\qquad
k = 30
$$

na célula abaixo.

In [5]:
A = np.array([[2,-2,-4],[-1,3,4],[1,-2,-3]])

print(matpot(A, k=30))




[[ 2 -2 -4]
 [-1  3  4]
 [ 1 -2 -3]]


(b) Você observou algo interessante no resultado?

Sim, percebe-se que ao modificar o k arbitráriamente ou defini-lo como k=30 solicitada pela questão esta matriz em específico conserva todos os seus coeficientes da matriz "original". Neste caso, $A^{n}=A, \forall n \in \mathbb{R}$.

(c) Agora, implemente o algoritmo matricial de Fibonacci utilizando a função `matpot` definida acima.
A função `fibmat` deve retornar o valor de $F_n$.

In [55]:
def fibmat(n):
    """
    Algoritmo matricial para o cálculo dos números da sequência de Fibonacci.
    
    Entrada:
         n (int): número inteiro não negativo.
         
    Saída:
         (int): o valor de F_n
    """
    # você deve iniciar a implementação desta função a partir daqui.
    assert n >= 0
    
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        M = np.array([[1,1],[1,0]])
        Mn = M
        for i in range(n):
          Mn = np.dot(M, Mn)
        return Mn[1,1]


Se tudo estiver OK com sua implementação, o resultado da execução da célula abaixo deverá ser `True`.

In [56]:
%%time
fibmat(32) == 2178309


CPU times: user 127 µs, sys: 0 ns, total: 127 µs
Wall time: 131 µs


True

(d) Você percebeu alguma diferença no tempo total de CPU desse método quando comparado ao método recursivo? Caso positivo, qual deles foi mais rápido? Você teria uma justificativa para isto? Discuta com seus amigos seus resultados.

Sim. O método matricial, assumindo $n=32$ retornou o $F_{n}$ com 131 $\mu s$ comparado ao método recursivo que retornou $F_{n}$ em um tempo de $1.24 s$.
Com o auxílio da biblioteca Numpy e com o uso da função $numpy.dot()=\sum\left [ A*B \right ]$ e usando recursividade juntamente com os critérios matemáticos estabelecidos anteriormente para obter o n-ésimo número de Fibonacci é possível obter $F_{n}$ de maneira mais eficiente que utilizando o método de recursividade tradicional.

## Saiba mais

- O problema de Fibonacci abordado no final da aula foi inspirado nas Miniaturas 1 e 2 do livro do professor Jiří Matoušek,  [Thirty-three Miniatures: Mathematical and Algorithmic Appplications of Linear Algebra](https://kam.mff.cuni.cz/~matousek/stml-53-matousek-1.pdf), *Student Mathematical Library*, v. 53, AMS, 2010.


* Ficou curioso e deseja conhecer outros algoritmos para o cálculo dos números de Fibonacci? Dê uma olhada no artigo: 
DASDAN, Ali. [Twelve Simple Algorithms to Compute Fibonacci Numbers](https://arxiv.org/abs/1803.07199). *arXiv preprint: 1803.07199*, 2018.

&copy; 2021 Vicente Helano  
UFCA | Centro de Ciências e Tecnologia